clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +256,465 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+            } else {
+                // Member is not part of the group anymore. Invoke onPartitionsLost.
+                callbackResult = invokeOnPartitionsLostCallback(droppedPartitions);
+            }
+        }
+        return callbackResult;
+    }
+
+    /**
+     * Reset member epoch to the value required for the leave the group heartbeat request, and
+     * transition to the {@link MemberState#SENDING_LEAVE_REQUEST} state so that a heartbeat
+     * request is sent out with it.
+     */
+    private void transitionToSendingLeaveGroup() {
+        memberEpoch = leaveGroupEpoch();
+        currentAssignment = new HashSet<>();
+        targetAssignment = Optional.empty();
+        transitionTo(MemberState.SENDING_LEAVE_REQUEST);
+    }
+
+    /**
+     * Return the epoch to use in the Heartbeat request to indicate that the member wants to
+     * leave the group. Should be -2 if this is a static member, or -1 in any other case.
+     */
+    private int leaveGroupEpoch() {
+        return groupInstanceId.isPresent() ? LEAVE_GROUP_EPOCH_FOR_STATIC_MEMBER : LEAVE_GROUP_EPOCH;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean shouldHeartbeatNow() {
+        return state() == MemberState.SENDING_ACK_FOR_RECONCILED_ASSIGNMENT ||
+                state() == MemberState.SENDING_LEAVE_REQUEST;
+    }
+
+    /**
+     * {@inheritDoc}
      */
-    private boolean maybeTransitionToStable() {
-        if (!hasPendingTargetAssignment()) {
+    @Override
+    public void onHeartbeatRequestSent() {
+        if (state() == MemberState.SENDING_ACK_FOR_RECONCILED_ASSIGNMENT) {
             transitionTo(MemberState.STABLE);
+        } else if (state() == MemberState.SENDING_LEAVE_REQUEST) {
+            transitionTo(MemberState.NOT_IN_GROUP);
+        }
+    }
+
+    @Override
+    public boolean shouldSkipHeartbeat() {
+        return state() == MemberState.NOT_IN_GROUP || state() == MemberState.FATAL;
+    }
+
+    void reconcile(ConsumerGroupHeartbeatResponseData.Assignment targetAssignment) {
+
+        SortedSet<TopicPartition> ownedPartitions = new TreeSet<>(COMPARATOR);
+        ownedPartitions.addAll(subscriptions.assignedPartitions());
+
+        CompletableFuture<SortedSet<TopicPartition>> assignedPartitionsByNameResult = extractTopicPartitionsFromAssignment(targetAssignment);
+
+        assignedPartitionsByNameResult.whenComplete((assignedPartitions, metadataError) -> {
+            if (metadataError != null) {
+                log.error("Reconciliation failed due to error getting metadata to resolve topic " +
+                        "names for topic IDs {} in target assignment.", targetAssignment);
+                // TODO: failing reconciliation (no ack sent to the broker), but leaving
+                //  member in STABLE state. Double check if any other action should be taken
+                //  here.
+                transitionTo(MemberState.STABLE);
```

Comment:
@dajac I think we should have both points solved now with the new metadata approach.

  1. We continue to request metadata updates as long as there are assigned topic IDs not resolved ([here](https://github.com/apache/kafka/blob/1c571fb9ca419f2916d663eaae6bd06b972b90bc/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L878)). This will solve the first case as you described.

  2. We keep a local cache of assigned topicId->topicNames for assigned topics that have been previously resolved. If topic is not in metadata when it comes in a next target assignment it will be resolved from the local cache ([here](https://github.com/apache/kafka/blob/1c571fb9ca419f2916d663eaae6bd06b972b90bc/clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java#L609)), as it is a known/assigned topic. This will solve case 2.
  
Thoughts?

Reply:
@lianetm I think that a client could still get an unresolvable topic id and be stuck in the reconciling state. For instance, it could happen if the member sees a topic id for the first time and the topic id is deleted just before it could resolve it.

Reply:
Agree. Just for the record, we're trying to figure out how to properly handle the cases where metadata wouldn't be available for a target assignment (permanently when topic deleted, or even just temporarily). With the current shape where the client attempts to reconcile the full target assignment, what happens now is rather disruptive, as the member would be kicked out of the group after rebalance timeout expires.